extension views API

sqs opened the pull request:
Adds a new (experimental) extension API `sourcegraph.app.registerViewProvider` and `ViewProvider` that allows extensions to display pages (with Markdown content and scoped search bars) on the Sourcegraph instance at the URL path `/views/ID` (where `ID` is the view ID chosen by the extension).

This is useful for:

- Displaying content that is globally relevant (such as global statistics), not per-repository
- In the future, creating purpose-built forms (by extending the ViewProvider to support more components, such as a JSON Schema-based form generator with onChange and onSubmit callbacks to registered extension commands)
- In the future, extending the views API to show partial views in more places (e.g., repository pages)

See https://github.com/sourcegraph/sourcegraph-custom-views for an example extension that uses this, and https://sourcegraph.slack.com/archives/C0B2RU51Q/p1587197734394700 for a screenshot of what it looks like. See https://docs.google.com/presentation/d/1FQ03zih1vlxCvAJyivLxHnvMXpN7fz0VF-K-QQd1fWk/edit#slide=id.g7d3a321198_0_311 for a screenshot of what a purpose-built form could look like (this PR does not support those yet, since that requires a JSON Schema form generator and change/submit callbacks, but it gets us most of the way there).

sqs commented:
@felixfbecker Can you or any other web team folks review this in the next couple days?

sqs commented on shared/src/schema/extension.schema.json:
```diff
@@ -203,6 +203,28 @@
+            "required": ["id", "where"],
+            "properties": {
+              "id": {
+                "description": "The identifier for this view, which must be unique among all contributed views. By convention, this is a dotted string of the form `myExtensionName.myViewName`.",
```
Yes, for now. To improve this in the future, I think we should add a URL alias that if specified takes precedence for URL routing. I think it's good to keep an ID that is more descriptive, for debugging and in-code referenceability.

sqs commented on packages/sourcegraph-extension-api/src/sourcegraph.d.ts:
```diff
@@ -697,6 +697,40 @@ declare module 'sourcegraph' {
+         */
+        content: (
+            | MarkupContent
+            | { component: string; props: { [name: string]: string | number | boolean | null | undefined } }
```
Yep, an `enum` is definitely good in the future. For now, given it's experimental, I think it's OK to merge as-is.

sqs replied:
I also would like to avoid an enum because that actually could imply more certainty/stability than just a string. (For now.)

sqs commented:
@eseliger Thank you for the review! Will go address your feedback in the code.

Re: rendering HTML, yes, definitely. We can do that either with a new `MarkupKind` or new `component`s.

felixfbecker commented:
I'll look at this in depth today or tomorrow.

Since this overlaps a lot with code insights I would like to avoid a situation where I don't know what the consequences are of me breaking something here. Could you:
- Confirm that this will _only_ be used on sourcegraph.com? (so we don't have to worry about old instances)
- Give the web team full access (and awareness) of any extension that uses this API? (so we can make breaking changes to the API as long as we update the extension)

felixfbecker commented:
I rebased on latest master and fixed merge conflicts from the comlink upgrade

sqs commented on shared/src/schema/extension.schema.json:
```diff
@@ -203,6 +203,28 @@
+              "where": {
+                "description": "Where the view is displayed. Currently the only supported value is 'global' (for views displayed as standalone pages at /views/ID).",
+                "type": "string",
+                "enum": ["global"]
```
Changed to `global/page` (the `/` notation matches that for menus). I think using the term `global` instead of `dedicated` is good because I anticipate having other possible values `user/page`, `organization/page`, `repository/page`, etc., so `global` is "parallel" conceptually to those.

sqs commented:
@felixfbecker Thank you. I just addressed all of your comments. PTAL.

> Confirm that this will only be used on sourcegraph.com? (so we don't have to worry about old instances)

Yes

> Give the web team full access (and awareness) of any extension that uses this API? (so we can make breaking changes to the API as long as we update the extension)

Yes

You OK with merging it now? I would love to be able to prototype some new pages on Sourcegraph.com with this.

For the `{component, props}` API, I was also thinking about JSX-like stuff over Comlink (based on what you told me a couple weeks ago). I agree it's not a nice API now. You OK with merging that not-nice API (marked experimental) for now? You can change and improve it however you want going forward.

felixfbecker commented:
Awesome! This is all I wanted to hear:

> You can change and improve it however you want going forward.

Looks like tests are failing in CI though.

felixfbecker commented on shared/src/api/extension/api/views.ts:
```diff
@@ -59,13 +62,22 @@ class ExtPanelView implements sourcegraph.PanelView {
+        const providerFunction: comlink.Local<
+            Parameters<ClientViewsAPI['$registerViewProvider']>[1]
+        > = comlink.proxy((params: { [key: string]: string }) =>
+            toProxyableSubscribable(provider.provideView(params), view => view!)
```
This `view!` non-null assertion looks incorrect since providers are allowed to emit `null`, and the code on the main thread does not handle `null` (you get type errors when removing this assertion that look legit to me)

felixfbecker commented:
I have two more questions (not blocking merging):
- Was the choice to put this into `enterprise/` intentional? I'd like to evolve these APIs into code insights, so I'd need to know if I should build the entirety of code insights (which contribute to various non-enterprise pages) as an enterprise-only feature, or if I should move the code into non-enterprise when I do that (in which case we might as well move it there right now to prevent hard-to-review diffs)
- Do you foresee other use cases of "custom components" atm? If so, I'd like to consider them when designing code insights APIs

sqs commented:
@felixfbecker:

> Was the choice to put this into enterprise/ intentional? I'd like to evolve these APIs into code insights, so I'd need to know if I should build the entirety of code insights (which contribute to various non-enterprise pages) as an enterprise-only feature, or if I should move the code into non-enterprise when I do that (in which case we might as well move it there right now to prevent hard-to-review diffs)

With the way this turned out, I don't think it needs to be in `enterprise/`. I will move it before merging. The extensibility points can be OSS; most or all of the specific code insights stuff will probably be non-OSS.

> Do you foresee other use cases of "custom components" atm? If so, I'd like to consider them when designing code insights APIs

Yes, a form that is autogenerated from a JSON Schema and where you can provide commands called on update and on submit. I have this in a branch on top of this branch, actually.

felixfbecker commented:
@sqs I think you force-pushed over my commits:
fcaef4c259b3f08b818be7550baefb5e1d2e431b
702616c85cca1e3df1815c1e8b1b1a933ced1a31
e8e363a5079bbfaac347f1a503963b3bdcc4d39d
a1bdf1145bd18791225517903fb1d9baa38bb42b

felixfbecker commented:
Opened a PR to add them back: https://github.com/sourcegraph/sourcegraph/pull/10186
